utilgo/tlsmanager/tlsmanager.go

```diff
@@ -0,0 +1,519 @@
+}
+
+const (
+	NamespaceDefault string = "default"
+	NamespaceArgoCD  string = "argocd"
+	CompanyName      string = "Atlas"
+	CountryName      string = "US"
+)
+
+type TLSSecretName string
+
+const (
+	NotValidSecretName            TLSSecretName = "not-valid"
+	WorkflowTriggerTLSSecretName  TLSSecretName = "workflowtrigger-tls"
+	ClientWrapperTLSSecretName    TLSSecretName = "clientwrapper-tls"
+	RepoServerTLSSecretName       TLSSecretName = "pipelinereposerver-tls"
+	CommandDelegatorTLSSecretName TLSSecretName = "commanddelegator-tls"
+	ArgoCDRepoServerTLSSecretName TLSSecretName = "argocd-repo-server-tls"
+	KafkaTLSSecretName            TLSSecretName = "kafka-tls"
+)
+
+type ClientName string
+
+const (
+	ClientRepoServer       ClientName = "reposerver"
+	ClientWorkflowTrigger  ClientName = "workflowtrigger"
+	ClientClientWrapper    ClientName = "clientwrapper"
+	ClientCommandDelegator ClientName = "commanddelegator"
+	ClientArgoCDRepoServer ClientName = "argocdreposerver"
+	ClientKafka            ClientName = "kafka"
+)
+
+type DNSName string
+
+const (
+	NotValidDNSName     DNSName = "not-valid"
+	DNSLocalhost        DNSName = "localhost"
+	DNSRepoServer       DNSName = "reposerver.default.svc.cluster.local"
+	DNSWorkflowTrigger  DNSName = "workflowtrigger.default.svc.cluster.local"
+	DNSCommandDelegator DNSName = "commanddelegator.default.svc.cluster.local"
+	DNSArgoCDRepoServer DNSName = "argocd-server.argocd.svc.cluster.local"
+)
+
+const (
+	TLSSecretCrtName      = "tls.crt"
+	TLSSecretKeyName      = "tls.key"
+	TLSSecretKafkaCrtName = "kafka.cert.pem"
+)
+
+func New(k kclient.KubernetesClient) Manager {
+	m := &tlsManager{k: k}
+	m.tlsClientConfigs = make(map[ClientName]*tls.Config)
+	m.tlsClientCertPEM = make(map[ClientName][]byte)
+	return m
+}
+
+func (m *tlsManager) BestEffortSystemCertPool() *x509.CertPool {
+	rootCAs, _ := x509.SystemCertPool()
+	if rootCAs == nil {
+		return x509.NewCertPool()
+	}
+	return rootCAs
+}
+
+func (m *tlsManager) GetServerTLSConf(serverName ClientName) (*tls.Config, error) {
+	conf, err := m.getTLSConf(serverName)
+	if err != nil {
+		return nil, err
+	}
+	m.tlsConf = conf
+	return conf, nil
+}
+
+func (m *tlsManager) GetClientTLSConf(clientName ClientName) (*tls.Config, error) {
+	conf, err := m.getTLSClientConf(clientName)
+	if err != nil {
+		return nil, err
+	}
+	m.tlsClientConfigs[clientName] = conf
+	return conf, nil
+}
+
+func (m *tlsManager) GetClientCertPEM(clientName ClientName) ([]byte, error) {
+	if m.tlsClientCertPEM[clientName] != nil {
+		return m.tlsClientCertPEM[clientName], nil
+	}
+
+	secretName, err := clientNameToSecretName(clientName)
+	if err != nil {
+		return nil, err
+	}
+
+	secret := m.k.FetchSecretData(string(secretName), NamespaceDefault)
+	if secret == nil || len(secret.Data) == 0 {
+		return nil, nil
+	}
+	return secret.Data[TLSSecretCrtName], nil
+}
+
+func (m *tlsManager) GetKafkaTLSConf() (*tls.Config, error) {
+	if m.tlsClientConfigs[ClientKafka] != nil {
```

Comment:
Make sure protections are in place for unnecessary watch secret triggers

Reply:
Added checks to all watcher methods